XIPs/xip-31-message-backup-request-system.md

```diff
@@ -0,0 +1,466 @@
+- Would require mobile application developers to handle XMTP deep links for app -> app transfer. This gets complicated when there are multiple potential applications as backup providers. The Backup Requester would need to be aware of all possible Backup Provider apps and the user would have to select the right one.
+- Would need to be rethought for webapp -> server transfer
+- Would need to be rethought for webapp -> webapp transfers where QR code scanning and deep-linking are not possible. Maybe something closer to oAuth.
+
+#### Trusted Backup Service
+
+This approach is similar to the Backup Account File, but the backup account would be run on a server and listen for new messages continuously. Messages could be added to an archive in real-time, and backups could be served to any application that provided appropriate authentication over HTTP or XMTP protocols.
+
+##### Trusted Service Advantages
+
+- Continuous real-time backup of all messages after the service is created. Backups would continue even if all of a user's devices were lost or compromised.
+- No need to involve any other applications in backup process. User would have complete control over where their backups were stored
+- Users could self-host trusted services
+
+##### Trusted Service Disadvantages
+
+- The service provider could read all messages as they come off the network. Even if they were encrypted at rest, there is significant risk of compromise.
+- More costly to run than the Backup Account File, since it requires server resources to be continuously polling/streaming from the network.
+- Multi-tenant providers could be legally compelled to provide access to user's messaging history or otherwise spy on users.
+- Self hosting with a high level of security would be a high bar and might be prohibitive for all but the most committed users.
+- Nothing is stopping users from doing this under any scenario. The trusted server would just be a client. If there is demand for this, someone will build it regardless.
+
+## Backward compatibility
+
+Given that V3 is brand new, there is no risk of backwards incompatible changes with the initial release of this feature.
+
+Caution should be taken in the design of the Message Backup File format to ensure that it is flexible and compatible with other potential changes. Any fields that may have changing data formats should be self-describing and versioned for backwards/forwards compatibility.
+
+## Reference implementation
+
+TODO
+
+## Security considerations
+
+### Risks and drawbacks to Remote Message Backups
+
+- Developers of the Message Backup Provider could implement the consent flow in a way that confuses users into accepting backup requests from malicious applications. PIN verification may not be implemented across all providers
+- A malicious Message Backup Provider could choose to censor messages in the backup, or add spoofed messages, since the messages in the backup are unauthenticated.
```

Comment:
If we want to add defense in depth:

1) We could consider requiring a wallet signature in the backup provider before sending the backup response. This allows the application layer to prove to the SDK layer that the correct information (including PIN) was surfaced to the user and consent was provided, minimizing implementation error, and requiring less context from app developers. In the current design, the security of remote backups is only as strong as the weakest app in the system.

2) We could consider requiring a wallet signature in the backup requester (optionally rolled into the same signature for installation key provisioning), to be validated on the provider before surfacing the consent flow. A particular area of potential weakness is the behavior with revoked installations. With the current design, any bug or momentary lapse in revocation logic (for example if a revocation payload was hidden from an installation for a brief moment in time) could allow a revoked installation to issue backup requests, or allow backup spoofing from a revoked installation. Additionally, in some cases a user may not actually want their message history on a new app, but there is nothing stopping a backup request from being issued regardless.

Arguable if these are worth the extra friction, mentioning here to provoke discussion

Reply:
It would also be nice to restrict the situations in which a backup can be requested - ideally only when creating a new installation, and at no other time.

Reply:
> We could consider requiring a wallet signature in the backup provider before sending the backup response. This allows the application layer to prove to the SDK layer that the correct information (including PIN) was surfaced to the user and consent was provided

While I guess it would lower the risk of developer error, the backup provider by definition already has all the messages. Why would they go through the trouble of using the backup service to exfiltrate messages without user consent when they can just use the SDK directly to get them? 

> A particular area of potential weakness is the behavior with revoked installations. With the current design, any bug or momentary lapse in revocation logic (for example if a revocation payload was hidden from an installation for a brief moment in time) could allow a revoked installation to issue backup requests, or allow backup spoofing from a revoked installation.

I'd like to go deeper on what the concerns with revocation are. It seems like we should address those head-on before we make any decisions that try to compensate for perceived weaknesses. Having access to the installation keys should be a strong assurance that the installation is controlled by the wallet owner given we are trusting this for message authentication. As part of the implementation we can (and should) force the Backup Provider to refresh the wallet address's contact topic to see if there are any recent revocation messages.

> Additionally, in some cases a user may not actually want their message history on a new app, but there is nothing stopping a backup request from being issued regardless.

I wonder if we can handle this point through strong UX guidance, rather than relying on wallet signatures. I worry that if we have too many wallet signatures as part of sign-in we will induce signature-blindness in the user, risking them clicking through signature dialogs carelessly. I do see the benefit in making backup requests an explicit user action, rather than an implicit part of sign-up. It could be annoying to receive a bunch of push notifications any time you access an XMTP app. Especially if apps don't explain what is happening to the user clearly.

> It would also be nice to restrict the situations in which a backup can be requested - ideally only when creating a new installation, and at no other time.

I like this. We should figure out how to enforce it, but it's a sensible idea.

Reply:
> Why would they go through the trouble of using the backup service to exfiltrate messages without user consent when they can just use the SDK directly to get them?

To clarify, the goal is not to protect against malicious backup providers, just careless ones which may then serve backups to a malicious requester.

> We should figure out how to enforce it, but it's a sensible idea.

A possible idea is to include this as part of the signing step for a new installation. For example, signing step looks like this:

```
XMTP: Grant Messaging Access

Current Time: <current time and local timezone>
Installation Key: <Curve25519PublicKey:Ed25519PublicKey>

Permissions:
- Full message history
```

When a provider receives a backup request, they can look for this signature, verify that 'full message history' has been granted, and only serve backups if the local clock is within 10 mins of what the user signed for.